--- Day 6: Trash Compactor ---
After helping the Elves in the kitchen, you were taking a break and helping them re-enact a movie scene when you over-enthusiastically jumped into the garbage chute!

A brief fall later, you find yourself in a garbage smasher. Unfortunately, the door's been magnetically sealed.

As you try to find a way out, you are approached by a family of cephalopods! They're pretty sure they can get the door open, but it will take some time. While you wait, they're curious if you can help the youngest cephalopod with her math homework.

In [1]:
elf_data = []

with open('day06.txt', 'r') as file:
    for line in file:
        elf_data.append(line.replace("\n", ""))

print(len(elf_data))
print(elf_data[0][0:15])
print(elf_data[1][0:15])
print(type(elf_data))


5
 8 797  34  5 8
75 142 794 58 4
<class 'list'>


Cephalopod math doesn't look that different from normal math. The math worksheet (your puzzle input) consists of a list of problems; each problem has a group of numbers that need to either be either added (+) or multiplied (*) together.

However, the problems are arranged a little strangely; they seem to be presented next to each other in a very long horizontal list. For example:

123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  

Each problem's numbers are arranged vertically; at the bottom of the problem is the symbol for the operation that needs to be performed. Problems are separated by a full column of only spaces. The left/right alignment of numbers within each problem can be ignored.

So, this worksheet contains four problems:

123 * 45 * 6 = 33210
328 + 64 + 98 = 490
51 * 387 * 215 = 4243455
64 + 23 + 314 = 401

In [2]:
test_data = [
"123 328  51 64 ", 
" 45 64  387 23 ", 
"  6 98  215 314",
"*   +   *   +  "]

test_data

['123 328  51 64 ', ' 45 64  387 23 ', '  6 98  215 314', '*   +   *   +  ']

In [3]:
import numpy as np
import regex as re

In [4]:
test_data1 = []
for i in range(0, len(test_data)):
    td=re.split(r' +', test_data[i])
    # hacky hack
    if td[-1] == '':
        td = td[:-1]
    if td[0] == '':
        td = td[1:]    
    test_data1.append(td)
test_data1

[['123', '328', '51', '64'],
 ['45', '64', '387', '23'],
 ['6', '98', '215', '314'],
 ['*', '+', '*', '+']]

In [5]:
elf_data1 = []
for i in range(0, len(elf_data)):
    ed = re.split(r' +', elf_data[i])
    #hacky hack
    if ed[-1] == '':
        ed = ed[:-1]
    if ed[0] == '':
        ed = ed[1:] 
    elf_data1.append(ed)

#elf_data1[:-1]

In [6]:
np_test_data1 = np.array(test_data1)
np_elf_data1 = np.array(elf_data1)
np_test_data1

array([['123', '328', '51', '64'],
       ['45', '64', '387', '23'],
       ['6', '98', '215', '314'],
       ['*', '+', '*', '+']], dtype='<U3')

In [7]:
tdt = np.transpose(np_test_data1)
edt = np.transpose(np_elf_data1)
tdt

array([['123', '45', '6', '*'],
       ['328', '64', '98', '+'],
       ['51', '387', '215', '*'],
       ['64', '23', '314', '+']], dtype='<U3')

To check their work, cephalopod students are given the grand total of adding together all of the answers to the individual problems. In this worksheet, the grand total is 33210 + 490 + 4243455 + 401 = 4277556.

In [8]:
def calculate(data):
    total = 0
    for d in data:
        #print(d)
        if d[-1] == "+":
            total += adding(d[:-1])
        else:
            if d[-1] == "*":
                total += multi(d[:-1])
        #print(d, total)
    return total

def adding(numbers):
    total = 0
    for n in numbers:
        total = total+int(n)
    print(total, "adding ", numbers)        
    return total

def multi(numbers):
    total = 1
    for n in numbers:
        total = total * int(n)
    print(total, "multip ", numbers)
    return total

calculate(tdt)

33210 multip  ['123' '45' '6']
490 adding  ['328' '64' '98']
4243455 multip  ['51' '387' '215']
401 adding  ['64' '23' '314']


4277556

In [9]:
tdt[:-1]

array([['123', '45', '6', '*'],
       ['328', '64', '98', '+'],
       ['51', '387', '215', '*']], dtype='<U3')

Of course, the actual worksheet is much wider. You'll need to make sure to unroll it completely so that you can read the problems clearly.

Solve the problems on the math worksheet. What is the grand total found by adding together all of the answers to the individual problems?


In [10]:
calculate(edt)

400200 multip  ['8' '75' '23' '29']
1355 adding  ['797' '142' '395' '21']
1742 adding  ['34' '794' '753' '161']
1608340 multip  ['5' '58' '59' '94']
257 adding  ['84' '45' '42' '86']
156 adding  ['14' '55' '22' '65']
2080 adding  ['57' '811' '979' '233']
733891185 multip  ['9' '965' '123' '687']
116961588 multip  ['831' '634' '37' '6']
20718 adding  ['5934' '9431' '5319' '34']
58979053971 multip  ['639' '233' '479' '827']
12460989648 multip  ['888' '478' '947' '31']
78 adding  ['9' '1' '1' '67']
548352 multip  ['12' '16' '34' '84']
6532 adding  ['6421' '66' '33' '12']
2497 adding  ['825' '877' '764' '31']
17235 adding  ['8418' '7192' '744' '881']
762 adding  ['9' '22' '32' '699']
23310 multip  ['6' '3' '37' '35']
1765218 multip  ['21' '53' '26' '61']
15160464306 multip  ['54' '741' '867' '437']
8247078 multip  ['37' '61' '58' '63']
1444 adding  ['95' '56' '576' '717']
1445 adding  ['138' '328' '887' '92']
8411 adding  ['17' '599' '2916' '4879']
144088892496 multip  ['778' '919' '216' '

5381996914800

--- Part Two ---
The big cephalopods come back to check on how things are going. When they see that your grand total doesn't match the one expected by the worksheet, they realize they forgot to explain how to read cephalopod math.

Cephalopod math is written right-to-left in columns. Each number is given in its own column, with the most significant digit at the top and the least significant digit at the bottom. (Problems are still separated with a column consisting only of spaces, and the symbol at the bottom of the problem is still the operator to use.)

Here's the example worksheet again:

123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
Reading the problems right-to-left one column at a time, the problems are now quite different:

In [11]:
def process_data(data):
    data2 = []
    l = max(len(d) for d in data)
    for d in data: 
        d2 = list(d)
        if len(d2) < l:
            for i in range(len(d2), l):
                d2.append('')
        #print(d2)
        #number = int(''.join(d2[:-1]))
        #data2.append(number,d2[-1])
        data2.append(d2)
    return np.array(np.transpose(data2))

td2 = process_data(test_data)
ed2 = process_data(elf_data)
print(td2)

[['1' ' ' ' ' '*']
 ['2' '4' ' ' ' ']
 ['3' '5' '6' ' ']
 [' ' ' ' ' ' ' ']
 ['3' '6' '9' '+']
 ['2' '4' '8' ' ']
 ['8' ' ' ' ' ' ']
 [' ' ' ' ' ' ' ']
 [' ' '3' '2' '*']
 ['5' '8' '1' ' ']
 ['1' '7' '5' ' ']
 [' ' ' ' ' ' ' ']
 ['6' '2' '3' '+']
 ['4' '3' '1' ' ']
 [' ' ' ' '4' ' ']]


In [12]:
def new_calc(data):
    total = 0
    i = 0
    numbers = []
    sign = ""
    while i<len(data):
        print(data[i])
        if data[i][-1] != " ":   # start new calc
            if len(numbers) != 0: # process last calc if it exists
                if sign =="+":
                    total += adding(numbers[:-1])
                else:
                    if sign == "*":
                        total += multi(numbers[:-1])
            numbers = []
            sign = data[i][-1]
        numbers.append(get_number(data[i][:-1]))            
        i += 1
    # hacky hack for last calc
    if sign =="+":
        total += adding(numbers)
    else:
        if sign == "*":
            total += multi(numbers)
    return total

def get_number(numbers):
    n = ''.join(numbers)
    #print(n)
    return n

new_calc(td2)

['1' ' ' ' ' '*']
['2' '4' ' ' ' ']
['3' '5' '6' ' ']
[' ' ' ' ' ' ' ']
['3' '6' '9' '+']
8544 multip  ['1  ', '24 ', '356']
['2' '4' '8' ' ']
['8' ' ' ' ' ' ']
[' ' ' ' ' ' ' ']
[' ' '3' '2' '*']
625 adding  ['369', '248', '8  ']
['5' '8' '1' ' ']
['1' '7' '5' ' ']
[' ' ' ' ' ' ' ']
['6' '2' '3' '+']
3253600 multip  [' 32', '581', '175']
['4' '3' '1' ' ']
[' ' ' ' '4' ' ']
1058 adding  ['623', '431', '  4']


3263827

The rightmost problem is 4 + 431 + 623 = 1058
The second problem from the right is 175 * 581 * 32 = 3253600
The third problem from the right is 8 + 248 + 369 = 625
Finally, the leftmost problem is 356 * 24 * 1 = 8544
Now, the grand total is 1058 + 3253600 + 625 + 8544 = 3263827.

Solve the problems on the math worksheet again. What is the grand total found by adding together all of the answers to the individual problems?

In [13]:
for e in elf_data:
    print(e[-10:])

for e in ed2[-10:]:
    print(e[-10:])

  4  95  3
4 5  93 55
2 57 92 98
3 61  7 78
  *  +  * 
[' ' '4' '2' '3' ' ']
[' ' ' ' ' ' ' ' ' ']
['4' '5' '5' '6' '*']
[' ' ' ' '7' '1' ' ']
[' ' ' ' ' ' ' ' ' ']
['9' '9' '9' ' ' '+']
['5' '3' '2' '7' ' ']
[' ' ' ' ' ' ' ' ' ']
[' ' '5' '9' '7' '*']
['3' '5' '8' '8' ' ']


In [14]:
new_calc(ed2)

[' ' '7' '2' '2' '*']
['8' '5' '3' '9' ' ']
[' ' ' ' ' ' ' ' ' ']
['7' '1' '3' ' ' '+']
6165158 multip  [' 722', '8539']
['9' '4' '9' '2' ' ']
['7' '2' '5' '1' ' ']
[' ' ' ' ' ' ' ' ' ']
[' ' '7' '7' '1' '+']
17456 adding  ['713 ', '9492', '7251']
['3' '9' '5' '6' ' ']
['4' '4' '3' '1' ' ']
[' ' ' ' ' ' ' ' ' ']
[' ' '5' '5' '9' '*']
9158 adding  [' 771', '3956', '4431']
['5' '8' '9' '4' ' ']
[' ' ' ' ' ' ' ' ' ']
['8' '4' '4' '8' '+']
3294746 multip  [' 559', '5894']
['4' '5' '2' '6' ' ']
[' ' ' ' ' ' ' ' ' ']
['1' '5' '2' '6' '+']
12974 adding  ['8448', '4526']
['4' '5' '2' '5' ' ']
[' ' ' ' ' ' ' ' ' ']
[' ' '8' '9' '2' '+']
6051 adding  ['1526', '4525']
['5' '1' '7' '3' ' ']
['7' '1' '9' '3' ' ']
[' ' ' ' ' ' ' ' ' ']
['9' '9' '1' '6' '*']
13258 adding  [' 892', '5173', '7193']
[' ' '6' '2' '8' ' ']
[' ' '5' '3' '7' ' ']
[' ' ' ' ' ' ' ' ' ']
['8' '6' ' ' ' ' '*']
3344032176 multip  ['9916', ' 628', ' 537']
['3' '3' '3' ' ' ' ']
['1' '4' '7' '6' ' ']
[' ' ' ' ' ' ' ' ' ']
['5' '9' 

9627174150897

9627172009458 too low